src/main/java/at/helpch/placeholderapi/expansion/server/util/VersionHelper.java

```diff
@@ -0,0 +1,99 @@
+ * furnished to do so, subject to the following conditions:
+ * <p>
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * <p>
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package at.helpch.placeholderapi.expansion.server.util;
+
+import com.google.common.primitives.Ints;
+import org.bukkit.Bukkit;
+import org.bukkit.OfflinePlayer;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Class for detecting server version for legacy support :(
+ */
+public final class VersionHelper {
+
+    public static final String NMS_VERSION = getNmsVersion();
+
+    private static final int V1_17 = 1_17_0;
+
+    private static final int CURRENT_VERSION = getCurrentVersion();
+
+    private static final boolean IS_PAPER = checkPaper();
+
+    public static final boolean IS_1_17_OR_HIGHER = CURRENT_VERSION >= V1_17;
+
+    /**
+     * Check if the server has access to the Paper API
+     * Taken from <a href="https://github.com/PaperMC/PaperLib">PaperLib</a>
+     *
+     * @return True if on Paper server (or forks), false anything else
+     */
+    private static boolean checkPaper() {
+        try {
+            Class.forName("com.destroystokyo.paper.PaperConfig");
+            return true;
+        } catch (ClassNotFoundException ignored) {
+            return false;
+        }
+    }
+
+    /**
+     * Gets the current server version
+     *
+     * @return A protocol like number representing the version, for example 1.16.5 - 1165
+     */
+    private static int getCurrentVersion() {
+        // No need to cache since will only run once
+        final Matcher matcher = Pattern.compile("(?<version>\\d+\\.\\d+)(?<patch>\\.\\d+)?").matcher(Bukkit.getBukkitVersion());
+
+        final StringBuilder stringBuilder = new StringBuilder();
+        if (matcher.find()) {
+            stringBuilder.append(matcher.group("version").replace(".", ""));
+            final String patch = matcher.group("patch");
+            if (patch == null) stringBuilder.append("0");
+            else stringBuilder.append(patch.replace(".", ""));
+        }
+
+        //noinspection UnstableApiUsage
+        final Integer version = Ints.tryParse(stringBuilder.toString());
+
+        // Should never fail
+        if (version == null) throw new IllegalArgumentException("Could not retrieve server version!");
+
+        return version;
+    }
```

Comment:
Is there any other use for `CURRENT_VERSION` outside the `IS_1_17_OR_HIGHER` part?

If not, then we could check for the existance of `Bukkit.getServer().getMinecraftVersion()` that was added in 2020, and if it does, get the version through it, remove dots and use that version.
And if it doesn't exist is the build from before 2020, so before ~1.16 itself.

Alternatively, just use `Bukkit.getServer().getBukkitVersion().split("-")[0]` like we do already somewhere else...

This entire thing just seems overkill for its only purpose.

Reply:
Currently this is only used for one check, yes, but in the future we might implement more placeholders that target a certain version. Plus, the code only runs once and then the number is cached.
The code turns 1.20 into 1200 and 1.12.2 into 1122. Without the extra 0 for 1.20, 1.12.2 would be *greater* than 1.20.

Reply:
Perhaps the usage of protocol versions here could be beneficial. They are always increase per protocol with only a few MC versions sharing the same protocol version (Usually ones with bug fixes).

A method for it would be available through the Unsafe class. Not sure since when it exists tho and if it is only in Paper o also in Spigot...